### PR TITLE
[Merged by Bors] - feat(RingTheory/LocalProperties): being projective can be checked locally on stalks

### PR DESCRIPTION
**Pull request by erdOne**
... for finitely presented modules

Written at the AIM algebraic geometry workshop

Co-authored-by: David Swinarski

---

- [x] depends on: #18264

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 1]
```diff
@@ -315,4 +315,13 @@ lemma Module.FinitePresentation.isLocalizedModule_map
...
     simpa using LinearMap.congr_fun e (f x)
 
+instance Module.FinitePresentation.isLocalizedModule_mapExtendScalars
+    {M' : Type*} [AddCommGroup M'] [Module R M'] (f : M →ₗ[R] M') [IsLocalizedModule S f]
+    {N' : Type*} [AddCommGroup N'] [Module R N'] (g : N →ₗ[R] N') [IsLocalizedModule S g]
+    (Rₛ) [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ M'] [Module Rₛ N']
+    [IsScalarTower R Rₛ M'] [IsScalarTower R Rₛ N'] [IsLocalization S Rₛ]
+    [Module.FinitePresentation R M] :
+      IsLocalizedModule S (IsLocalizedModule.mapExtendScalars S f g Rₛ) :=
+  IsLocalizedModule.of_linearEquiv _ _ _
```
Comment (chrisflav) [inline Mathlib/Algebra/Module/FinitePresentation.lean:325]: This declaration shares most of its variables with the one above, can you make them into one section?

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+Authors: Andrew Yang
+-/
+import Mathlib.Algebra.Module.FinitePresentation
+import Mathlib.Algebra.Module.Projective
+import Mathlib.RingTheory.LocalProperties.Submodule
+import Mathlib.RingTheory.Localization.BaseChange
+
+/-!
+# Local properties of modules
+
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:13]:
```suggestion
# Local properties of projective modules

```
copy paste leftover?

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  then `M` being projective can be checked on stalks (of maximal ideals).
+- `Module.projective_of_localization_maximal` If `M` is finitely presented,
+  then `M` being projective can be checked on stalks (of maximal ideals).
+
+-/
+
+variable {R M N N'} [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+variable [AddCommGroup N'] [Module R N'] (S : Submonoid R)
+
+theorem split_surjective_of_localization_maximal (f : M →ₗ[R] N) [Module.FinitePresentation R N]
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:27]: Should this be in the `Module` namespace?

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+/-!
+# Local properties of modules
+
+## Main results
+- `split_surjective_of_localization_maximal` If `N` is finitely presented, then `f : M →ₗ[R] N`
+  being split injective can be checked on stalks (of maximal ideals).
+- `Module.projective_of_localization_maximal` If `M` is finitely presented,
+  then `M` being projective can be checked on stalks (of maximal ideals).
+- `Module.projective_of_localization_maximal` If `M` is finitely presented,
+  then `M` being projective can be checked on stalks (of maximal ideals).
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:20]: This is listed twice, did you mean to include something else here?

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    (LocalizedModule.map_surjective _ _ hf)
+  obtain ⟨g, hg⟩ := split_surjective_of_localization_maximal _ this
+  exact Module.Projective.of_split _ _ hg
+
+theorem Module.free_of_isLocalizedModule (Rₛ) {Mₛ} [AddCommGroup Mₛ] [Module R Mₛ]
+    [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ Mₛ] [IsScalarTower R Rₛ Mₛ]
+    (S) (f : M →ₗ[R] Mₛ) [IsLocalization S Rₛ] [IsLocalizedModule S f] [Module.Free R M] :
+      Module.Free Rₛ Mₛ :=
+  Module.Free.of_basis
+    (Basis.ofIsLocalizedModule Rₛ S f (Module.Free.chooseBasis R M))
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:108]: Can this go to an earlier file?

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 6]
```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:4]: David Swinarski is listed as Co-Author of the PR, maybe then also add them here?

File: Mathlib/RingTheory/LocalProperties/Submodule.lean [hunk 7]
```diff
@@ -0,0 +1,97 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.Algebra.Module.Submodule.Localization
+import Mathlib.RingTheory.Localization.AtPrime
+
+/-!
+# Local properties of modules
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Submodule.lean:10]:
```suggestion
# Local properties of submodules
```

File: Mathlib/RingTheory/LocalProperties/Submodule.lean [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+In this file, we show that several conditions on submodules can be checked on stalks.
+-/
+
+
+section Submodule
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Submodule.lean:16]:
```suggestion
```
The `section` does not seem to do anything

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      apply LinearMap.restrictScalars_injective R
+      apply IsLocalizedModule.ringHom_ext I.primeCompl (LocalizedModule.mkLinearMap I.primeCompl N)
+      · exact IsLocalizedModule.map_units (LocalizedModule.mkLinearMap I.primeCompl N)
+      ext; simp only [LocalizedModule.map, LinearMap.coe_comp,
+        LinearMap.coe_restrictScalars, Function.comp_apply, LocalizedModule.mkLinearMap_apply,
+        LocalizedModule.map_mk, LinearMap.smul_apply, LinearMap.llcomp_apply]
+
+theorem Module.projective_of_localization_maximal (H : ∀ (I : Ideal R) (_ : I.IsMaximal),
+    Module.Projective (Localization.AtPrime I) (LocalizedModule I.primeCompl M))
+    [Module.FinitePresentation R M] : Module.Projective R M := by
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:86]: Maybe also a general `IsLocalizedModule` version is useful?

File: Mathlib/RingTheory/LocalProperties/Submodule.lean [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  (Mₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module R (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module (Rₚ P) (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsScalarTower R (Rₚ P) (Mₚ P)]
+  (f : ∀ (P : Ideal R) [P.IsMaximal], M →ₗ[R] Mₚ P)
+  [inst : ∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (f P)]
+
+/-- Let `I J : Ideal R`. If the localization of `I` at each maximal ideal `P` is included in
+the localization of `J` at `P`, then `I ≤ J`. -/
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Submodule.lean:36]: The docstrings in this file seem to all be about the version for `Ideal`s. For the sake of intuitive notation, could you also replace `I` and `J` (which allude to ideals) by e.g. `N` and `O`?

**Comment by erdOne** [comment]
Thanks for the review!

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+import Mathlib.RingTheory.LocalProperties.Submodule
+import Mathlib.RingTheory.Localization.BaseChange
+
+/-!
+
+# Being projective is a local property
+
+## Main results
+- `split_surjective_of_localization_maximal` If `N` is finitely presented, then `f : M →ₗ[R] N`
+  being split injective can be checked on stalks (of maximal ideals).
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:18]:
```suggestion
- `LinearMap.split_surjective_of_localization_maximal` If `N` is finitely presented, then `f : M →ₗ[R] N`
  being split injective can be checked on stalks (of maximal ideals).
```

File: Mathlib/RingTheory/LocalProperties/Submodule.lean [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+/-- Let `N₁ N₂ : Submodule R M`. If the localization of `N₁` at each maximal ideal `P` is equal to
+the localization of `N₂` at `P`, then `N₁ = N₂`. -/
+theorem Submodule.eq_of_localization_maximal {N₁ N₂ : Submodule R M}
+    (h : ∀ (P : Ideal R) [P.IsMaximal],
+      N₁.localized' (Rₚ P) P.primeCompl (f P) = N₂.localized' (Rₚ P) P.primeCompl (f P)) :
+    N₁ = N₂ :=
+  le_antisymm (Submodule.le_of_localization_maximal Rₚ Mₚ f fun P _ => (h P).le)
+    (Submodule.le_of_localization_maximal Rₚ Mₚ f fun P _ => (h P).ge)
+
+/-- An ideal is trivial if its localization at every maximal ideal is trivial. -/
```
Comment (chrisflav) [inline Mathlib/RingTheory/LocalProperties/Submodule.lean:63]:
```suggestion
/-- A submodule is trivial if its localization at every maximal ideal is trivial. -/
```

**Review by chrisflav — APPROVED** [review]

**Comment by chrisflav** [comment]
The long line linter complains, otherwise I am happy.

**Comment by chrisflav** [comment]
Thanks!

maintainer merge

**Comment by erdOne** [comment]
Okay I think I have a better approach. Let's hold this for now.

**Comment by erdOne** [comment]
No I was wrong :((
Let's keep this open.

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  [∀ (P : Ideal R) [P.IsMaximal], CommRing (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Algebra R (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsLocalization.AtPrime (Rₚ P) P]
+  (Mₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module R (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module (Rₚ P) (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsScalarTower R (Rₚ P) (Mₚ P)]
+  (f : ∀ (P : Ideal R) [P.IsMaximal], M →ₗ[R] Mₚ P)
+  [inst : ∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (f P)]
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:125]: At some point we might want to have some syntax for variable blocks like this...

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+theorem Module.projective_of_isLocalizedModule {Rₛ Mₛ} [AddCommGroup Mₛ] [Module R Mₛ]
+    [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ Mₛ] [IsScalarTower R Rₛ Mₛ]
+    (S) (f : M →ₗ[R] Mₛ) [IsLocalization S Rₛ] [IsLocalizedModule S f] [Module.Projective R M] :
+      Module.Projective Rₛ Mₛ :=
+    Projective.of_equiv (IsLocalizedModule.isBaseChange S Rₛ f).equiv
+
+theorem LinearMap.split_surjective_of_localization_maximal
+    (f : M →ₗ[R] N) [Module.FinitePresentation R N]
+    (H : ∀ (I : Ideal R) (hI : I.IsMaximal),
```
Comment (Vierkantor) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:39]: The unused variable linter flagged this.
```suggestion
    (H : ∀ (I : Ideal R), I.IsMaximal →
```

**Review by Vierkantor — APPROVED** [review]
Could you please merge the `master` branch into this PR and then (assuming everything builds) reply with `bors r+`?

bors d+

**Comment by erdOne** [comment]
bors merge

**Pull request closed without merging**